### PR TITLE
OCPBUGS-30319: bump lib-go to fix SAs acting as OAuth2 clients

### PR DESCRIPTION
**Pull request by liouk**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,31 +5,31 @@ go 1.21
 require (
 	github.com/MakeNowJust/heredoc v1.0.0
 	github.com/google/btree v1.1.2
-	github.com/google/go-cmp v0.5.9
+	github.com/google/go-cmp v0.6.0
 	github.com/google/gofuzz v1.2.0
 	github.com/google/uuid v1.6.0
```
Comment (p0lyn0mial) [inline go.mod:10]: kube is on `v1.3.0` - https://github.com/kubernetes/kubernetes/blob/release-1.29/go.mod#L47

this dep was changed earlier, I think it is okay since minor version are compatible.

File: go.mod [hunk 2]
```diff
@@ -5,31 +5,31 @@ go 1.21
...
-	k8s.io/utils v0.0.0-20230505201702-9f6742963106
+	go.opentelemetry.io/otel/trace v1.19.0
+	k8s.io/api v0.29.2
+	k8s.io/apiextensions-apiserver v0.29.2
+	k8s.io/apimachinery v0.29.2
+	k8s.io/apiserver v0.29.2
+	k8s.io/client-go v0.29.2
+	k8s.io/code-generator v0.29.2
+	k8s.io/component-base v0.29.2
+	k8s.io/klog/v2 v2.120.1
```
Comment (p0lyn0mial) [inline go.mod:28]: kube is on `v2.110.1` - https://github.com/kubernetes/kubernetes/blob/release-1.29/go.mod#L114
let's match kube.

**Comment by p0lyn0mial** [comment]
Also verified if the termination events (https://github.com/openshift/kubernetes-apiserver/pull/56/commits/8ba4d58edc945b26be81691800a7361d0e726f75) are applied https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_oauth-apiserver/108/pull-ci-openshift-oauth-apiserver-master-e2e-aws/1768660242603708416/artifacts/e2e-aws/gather-must-gather/artifacts/event-filter.html

File: pkg/cmd/oauth-apiserver/cmd.go [hunk 3]
```diff
@@ -49,7 +49,6 @@ func NewOAuthAPIServerOptions(out io.Writer) *OAuthAPIServerOptions {
 		TokenValidationOptions: tokenvalidationoptions.NewTokenValidationOptions(),
 		Output:                 out,
 	}
-	o.RecommendedOptions.Etcd.StorageConfig.Paging = true
```
Comment (p0lyn0mial) [inline pkg/cmd/oauth-apiserver/cmd.go:52]: why the change ?

File: pkg/cmd/oauth-apiserver/cmd.go [hunk 4]
```diff
@@ -133,7 +132,7 @@ func (o *OAuthAPIServerOptions) NewOAuthAPIServerConfig() (*apiserver.Config, er
 	}
 
 	serverConfig.GenericConfig.OpenAPIConfig = openapiconfig.DefaultOpenAPIConfig()
-	serverConfig.GenericConfig.OpenAPIV3Config = openapiconfig.DefaultOpenAPIConfig()
+	serverConfig.GenericConfig.OpenAPIV3Config = openapiconfig.DefaultOpenAPIV3Config()
```
Comment (p0lyn0mial) [inline pkg/cmd/oauth-apiserver/cmd.go:135]: why the change ?

File: go.mod [hunk 5]
```diff
@@ -5,31 +5,31 @@ go 1.21
...
 	github.com/MakeNowJust/heredoc v1.0.0
 	github.com/google/btree v1.1.2
-	github.com/google/go-cmp v0.5.9
+	github.com/google/go-cmp v0.6.0
 	github.com/google/gofuzz v1.2.0
 	github.com/google/uuid v1.6.0
 	github.com/jteeuwen/go-bindata v3.0.8-0.20151023091102-a0ff2567cfb7+incompatible
-	github.com/openshift/api v0.0.0-20231101013329-0d0d46454bb7
-	github.com/openshift/apiserver-library-go v0.0.0-20230621110634-a99412818848
+	github.com/openshift/api v0.0.0-20231218131639-7a5aa77cc72d
```
Comment (p0lyn0mial) [inline go.mod:12]: why not to pull much newer version ?

Comment (p0lyn0mial) [reply go.mod:12]: it looks like the same applies to the other `openshift/*` deps.

File: go.mod [hunk 6]
```diff
@@ -86,83 +86,81 @@ require (
...
-	k8s.io/kubernetes => github.com/openshift/kubernetes v0.0.0-20231102175044-c01bd450743e
-	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.28.3
-	k8s.io/metrics => k8s.io/metrics v0.28.3
-	k8s.io/mount-utils => k8s.io/mount-utils v0.28.3
-	k8s.io/pod-security-admission => k8s.io/pod-security-admission v0.28.3
-	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.28.3
+	k8s.io/api => k8s.io/api v0.29.2
+	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.29.2
+	k8s.io/apimachinery => k8s.io/apimachinery v0.29.2
+	k8s.io/apiserver => github.com/liouk/kubernetes-apiserver v0.0.0-20240313155529-ced3d910063d // points to openshift-apiserver-4.16-kubernetes-1.29.2
```
Comment (p0lyn0mial) [inline go.mod:141]: don't forget to point it to the real repo :)

File: go.mod [hunk 7]
```diff
@@ -86,83 +86,81 @@ require (
...
+	k8s.io/csi-translation-lib => k8s.io/csi-translation-lib v0.29.2
+	k8s.io/dynamic-resource-allocation => k8s.io/dynamic-resource-allocation v0.29.2
+	k8s.io/kms => k8s.io/kms v0.29.2
+	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.29.2
+	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.29.2
+	k8s.io/kube-proxy => k8s.io/kube-proxy v0.29.2
+	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.29.2
+	k8s.io/kubectl => k8s.io/kubectl v0.29.2
+	k8s.io/kubelet => k8s.io/kubelet v0.29.2
+	k8s.io/kubernetes => github.com/openshift/kubernetes v1.29.0-rc.1.0.20240306191509-a0beecc776d4
```
Comment (p0lyn0mial) [inline go.mod:160]: this should point to `k/k` repo, right ?

Comment (p0lyn0mial) [reply go.mod:160]: /cc @dgrisonnet

File: pkg/cmd/oauth-apiserver/cmd.go (hunk 4 above)
Comment (liouk) [reply pkg/cmd/oauth-apiserver/cmd.go:135]: `DefaultOpenAPIConfig()` returns `*common.Config` which is incompatible with `common.OpenAPIV3config` (the type of `serverConfig.GenericConfig.OpenAPIV3Config`).

**Comment by liouk** [comment]
Updated PR to point to `openshift/kubernetes-apiserver@openshift-apiserver-4.16-kubernetes-1.29.2`.

File: pkg/cmd/oauth-apiserver/openapiconfig/openapiconfig.go [hunk 8]
```diff
@@ -115,3 +116,101 @@ func DefaultOpenAPIConfig() *openapicommon.Config {
 		SecurityDefinitions: &securityDefinitions,
 	}
 }
+
+func DefaultOpenAPIV3Config() *openapicommon.OpenAPIV3Config {
+	defNamer := apiserverendpointsopenapi.NewDefinitionNamer(legacyscheme.Scheme, extensionsapiserver.Scheme, aggregatorscheme.Scheme)
+	return &openapicommon.OpenAPIV3Config{
+		Info: &spec.Info{
```
Comment (p0lyn0mial) [inline pkg/cmd/oauth-apiserver/openapiconfig/openapiconfig.go:123]: It looks like `spec.Info` could be shared between `DefaultOpenAPIV3Config` and `DefaultOpenAPIConfig`, could you refactor ?

File: pkg/cmd/oauth-apiserver/openapiconfig/openapiconfig.go [hunk 9]
```diff
@@ -115,3 +116,101 @@ func DefaultOpenAPIConfig() *openapicommon.Config {
 		SecurityDefinitions: &securityDefinitions,
 	}
 }
+
+func DefaultOpenAPIV3Config() *openapicommon.OpenAPIV3Config {
+	defNamer := apiserverendpointsopenapi.NewDefinitionNamer(legacyscheme.Scheme, extensionsapiserver.Scheme, aggregatorscheme.Scheme)
```
Comment (p0lyn0mial) [inline pkg/cmd/oauth-apiserver/openapiconfig/openapiconfig.go:121]: It looks like we could create a default config using ` genericapiserver.DefaultOpenAPIV3Config` and then overwrite:
- `spec.Info`
- `IgnorePrefixes`
- `SecuritySchemes`

WDYT ?

File: pkg/cmd/oauth-apiserver/cmd.go (hunk 3 above)
Comment (p0lyn0mial) [reply pkg/cmd/oauth-apiserver/cmd.go:52]: ok, the param was removed in https://github.com/kubernetes/kubernetes/pull/121390

File: go.mod [hunk 10]
```diff
@@ -5,31 +5,31 @@ go 1.21
...
+	k8s.io/apimachinery v0.29.2
+	k8s.io/apiserver v0.29.2
+	k8s.io/client-go v0.29.2
+	k8s.io/code-generator v0.29.2
+	k8s.io/component-base v0.29.2
+	k8s.io/klog/v2 v2.120.1
+	k8s.io/kube-aggregator v0.29.2
+	k8s.io/kube-openapi v0.0.0-20231010175941-2dd684a91f00
+	k8s.io/kubernetes v1.29.2
+	k8s.io/utils v0.0.0-20240310230437-4693a0247e57
```
Comment (p0lyn0mial) [inline go.mod:32]: kube is on `v0.0.0-20230726121419-3b25d923346b` - https://github.com/kubernetes/kubernetes/blob/release-1.29/go.mod#L129C15-L129C49

File: pkg/cmd/oauth-apiserver/openapiconfig/openapiconfig.go (hunk 9 above)
Comment (liouk) [reply pkg/cmd/oauth-apiserver/openapiconfig/openapiconfig.go:121]: Good idea, thanks for pointing out `genericapiserver.DefaultOpenAPIV3Config`!

**Comment by liouk** [comment]
/retitle OCPBUGS-30319: Bump openshift/kubernetes-apiserver to openshift-apiserver-4.16-kubernetes-1.29.2

File: pkg/cmd/oauth-apiserver/openapiconfig/openapiconfig.go (hunk 9 above)
Comment (p0lyn0mial) [reply pkg/cmd/oauth-apiserver/openapiconfig/openapiconfig.go:124]: OK, once this PR merges can you open a new one for using `genericapiserver.DefaultOpenAPIConfig()` in `DefaultOpenAPIConfig` ?

File: go.mod (hunk 2 above)
Comment (p0lyn0mial) [reply go.mod:28]: ping ?

**Comment by p0lyn0mial** [comment]
LGTM

just two additional comments, do you mind addressing https://github.com/openshift/oauth-apiserver/pull/108#discussion_r1559159515 and https://github.com/openshift/oauth-apiserver/pull/108#discussion_r1560827754 ?

File: go.mod (hunk 2 above)
Comment (liouk) [reply go.mod]: This must have been reverted back with some other module upgrade :man_facepalming: Fixed

**Comment by liouk** [comment]
Proof PR updated and can now be used to perform the actual bump.

/retitle OCPBUGS-30319: bump lib-go to fix SAs acting as OAuth2 clients
/hold cancel
